### PR TITLE
Enable support for print_every_n and early_stopping_rounds

### PR DESCRIPTION
**Pull request by dmitrijsc**
Hello

I have recently been asking about adding additional parameters to nfold_cv (https://github.com/dmlc/XGBoost.jl/issues/40) and now I have done it myself.  
I have also addressed the issue described here: https://github.com/dmlc/XGBoost.jl/issues/42

I have added both print_every_n and early_stopping_rounds which are useful when running cross validation.

**print_every_n** - print each n-th iteration evaluation messages when silent = 0. Default is 1 which means all messages are printed. This parameter is passed to the nfold_cv or xgboost.

**early_stopping_rounds** - if not set, the early stopping function is not triggered. If set to an integer k > 0, training with a validation set will stop if the performance doesn't improve for k rounds. This parameter is passed to the nfold_cv or xgboost.

That's my first PR :)

### CODE REVIEW

File: src/xgboost_wrapper_h.jl [hunk 1]
```diff
@@ -10,8 +10,8 @@ end
 macro xgboost(f, params...)
     return quote
         err = ccall(($f, _jl_libxgboost), Int64,
-                    ($((esc(i.args[2]) for i in params)...),),
-                    $((esc(i.args[1]) for i in params)...))
+                    ($((esc(i.args[3]) for i in params)...),),
```
Comment (Allardvm) [inline src/xgboost_wrapper_h.jl]: `i.args[end]` here

File: src/xgboost_wrapper_h.jl [hunk 2]
```diff
@@ -10,8 +10,8 @@ end
...
+                    $((esc(i.args[2]) for i in params)...))
```
Comment (Allardvm) [inline src/xgboost_wrapper_h.jl]: and `i.args[end - 1]` here, for compatibility with 0.5

File: src/xgboost_lib.jl [hunk 3]
```diff
@@ -294,33 +300,78 @@ function aggcv(rlist; show_stdv = true)
...
+
+    res
 end
 
 function nfold_cv(data, num_boost_round::Integer = 10, nfold::Integer = 3; label = Union{},
                   param=[], metrics=[], obj = Union{}, feval = Union{}, fpreproc = Union{},
                   show_stdv = true, seed::Integer = 0, kwargs...)
     dtrain = makeDMatrix(data, label)
-    results = String[]
+    results = []
```
Comment (Allardvm) [inline src/xgboost_lib.jl]: Won't this turn it into a Vector{Any}? Initializing it with `String[]` results in more efficient code if we only push strings into the vector, so that'd be preferable to me.

**Comment by Allardvm** [comment]
Thanks for the contribution! I've added a few suggestions to your code that you could like at while waiting for the maintainers to chime in.

File: src/xgboost_lib.jl (hunk 3 above)
Comment (dmitrijsc) [inline src/xgboost_lib.jl]: It might be not the most efficient way, but the reason I am having a vector of any is to actually split the results from a string to multiple columns. These way we will be be able to work with the results and stop when test error is at its minimum.

File: src/xgboost_wrapper_h.jl (hunk 1 above)
Comment (dmitrijsc) [inline src/xgboost_wrapper_h.jl]: Yes, that's definitely makes sense here.

**Comment by slundberg** [comment]
Thanks! As @Allardvm mentioned it would be good to get it working on Julia 0.5. (Currently it crashes on load)

**Comment by dmitrijsc** [comment]
Thanks for feedback. I will see what's the issue and make a fix.

**Comment by slundberg** [comment]
Thanks. It looks like the travis failure is due to an unrelated compilation issue on linux. My guess is the Julia Travis CI for linux image had a compiler change recently that causes issues. @Allardvm is this PR safe to merge without causing trouble for the refactor you were doing?

**Comment by dmitrijsc** [comment]
I just realised it is too early to merge it with master. I have to implement "maximize" property.

From R docs:
**maximize** - If custom objective and early_stopping_rounds are set, then this parameter must be set as well. When it is TRUE, it means the larger the evaluation score the better.

**Comment by Allardvm** [comment]
@slundberg Yes, this is safe to merge. The refactor sticks as close as possible to the Python package (as discussed a while ago) and uses callbacks for the functionality that this PR adds. This required a complete rewrite of all the functions that this PR changes, so any changes here won't affect the refactor.

@dmitrijsc It might safe you a bit of time to mimic the implementation of the "maximize" property in the refactor's callbacks: https://github.com/Allardvm/XGBoost.jl/blob/package-refactor/src/callbacks.jl. Feel free to copy-paste anything that might help.

**Comment by BabaYara** [comment]
Was early stopping ever implemented?

**Comment by ablaom** [comment]
> Was early stopping ever implemented?

Not sure, but an option is to use the MLJ interface which allows external control of any iterative model, including out-of-sample loss stopping criterion

https://alan-turing-institute.github.io/MLJ.jl/dev/controlling_iterative_models/

**Comment by ExpandingMan** [comment]
This is no longer valid as of the merging of #111.  It still might be nice to make periodic reporting more flexible and to make it easier to set up the behavior that this PR facilitates.

Note that as with #97 it is now much easier for users to achieve this through public API calls since it now exposes update methods.
